### PR TITLE
Test generator for output consistency checks (V2): better type information

### PR DESCRIPTION
**Pull request by nrainer-materialize**
### Motivation
Improvements for https://github.com/MaterializeInc/materialize/issues/18858.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: misc/python/materialize/output_consistency/input_data/params/number_operation_param.py [hunk 1]
```diff
@@ -37,3 +48,29 @@ def __init__(
...
+    def supports_type(self, data_type: DataType) -> bool:
+        if not isinstance(data_type, NumberDataType):
+            return False
+
+        if self.only_int_type and data_type.is_decimal:
+            return False
+
+        if self.no_int_type_larger_int4 and data_type.identifier in {
+            INT8_TYPE_IDENTIFIER,
+            UINT4_TYPE_IDENTIFIER,
```
Comment (def-) [inline misc/python/materialize/output_consistency/input_data/params/number_operation_param.py:65]: Is uint4 larger than int4? Is this about the MAX value that fits inside?

File: misc/python/materialize/output_consistency/input_data/params/number_operation_param.py [hunk 2]
```diff
@@ -24,6 +31,10 @@ def __init__(
         incompatibility_combinations: Optional[
             List[Set[ExpressionCharacteristics]]
         ] = None,
+        only_int_type: bool = False,
+        no_int_type_larger_int4: bool = False,
+        no_floating_point_type: bool = False,
+        no_unsigned_type: bool = False,
```
Comment (def-) [inline misc/python/materialize/output_consistency/input_data/params/number_operation_param.py:37]: I think I might have created a list of types and implemented these as a lambda/small function.

**Review by def- — CHANGES_REQUESTED** [review]

File: misc/python/materialize/output_consistency/input_data/params/number_operation_param.py (hunk 1 above)
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/input_data/params/number_operation_param.py:65]: Yes, it is about the max value. Any alternative naming suggestions?

File: misc/python/materialize/output_consistency/input_data/params/number_operation_param.py (hunk 2 above)
Comment (nrainer-materialize) [reply misc/python/materialize/output_consistency/input_data/params/number_operation_param.py:37]: Can you explain what the benefit would be?

File: misc/python/materialize/output_consistency/input_data/params/number_operation_param.py (hunk 1 above)
Comment (def-) [reply misc/python/materialize/output_consistency/input_data/params/number_operation_param.py:65]: Ok, my interpretation was that it uses more bytes, so uint4 and int4 would be the same size.
I don't have a good term, `type_fits_larger_values_than_INT4_MAX` seems a bit unwieldy.

File: misc/python/materialize/output_consistency/input_data/params/number_operation_param.py (hunk 2 above)
Comment (def-) [reply misc/python/materialize/output_consistency/input_data/params/number_operation_param.py:37]: The supports_type method could then be a trivial operation, and each user of NumericOperationParam could define their own filter. But I guess it's ok as is.

**Review by def- — APPROVED** [review]

**Pull request merged**